lib/core/src/Cardano/Wallet/DaedalusIPC.hs

```diff
@@ -0,0 +1,231 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Provides a mechanism for Daedalus to discover what port the cardano-wallet
+-- server is listening on.
+--
+-- See <https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options>
+-- for more information about the message protocol.
+
+module Cardano.Wallet.DaedalusIPC
+    ( daedalusIPC
+    ) where
+
+import Prelude
+
+import Control.Concurrent
+    ( threadDelay )
+import Control.Concurrent.Async
+    ( race_ )
+import Control.Concurrent.MVar
+    ( MVar, newEmptyMVar, putMVar, takeMVar )
+import Control.Exception
+    ( IOException, catch, tryJust )
+import Control.Monad
+    ( forever, when )
+import Data.Aeson
+    ( FromJSON (..)
+    , ToJSON (..)
+    , defaultOptions
+    , eitherDecode
+    , encode
+    , genericParseJSON
+    , genericToEncoding
+    )
+import Data.Aeson.Types
+    ( Options, SumEncoding (ObjectWithSingleField), sumEncoding )
+import Data.Bifunctor
+    ( first )
+import Data.Binary.Get
+    ( getWord32le, getWord64le, runGet )
+import Data.Binary.Put
+    ( putLazyByteString, putWord32le, putWord64le, runPut )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Text
+    ( Text )
+import Data.Word
+    ( Word32, Word64 )
+import GHC.Generics
+    ( Generic )
+import GHC.IO.Handle.FD
+    ( fdToHandle )
+import Say
+    ( sayErr, sayErrString )
+import System.Environment
+    ( lookupEnv )
+import System.Info
+    ( arch )
+import System.IO
+    ( Handle, hFlush, hGetLine, hSetNewlineMode, noNewlineTranslation, stdout )
+import System.IO.Error
+    ( IOError, isEOFError )
+import Text.Read
+    ( readEither )
+
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as L8
+import qualified Data.Text as T
+
+----------------------------------------------------------------------------
+-- Daedalus <-> Wallet child process port discovery protocol
+
+data MsgIn  = QueryPort
+    deriving (Show, Eq, Generic)
+data MsgOut = Started | ReplyPort Int | ParseError Text
+    deriving (Show, Eq, Generic)
+
+aesonOpts :: Options
+aesonOpts = defaultOptions { sumEncoding = ObjectWithSingleField }
+
+instance FromJSON MsgIn where
+    parseJSON = genericParseJSON aesonOpts
+instance ToJSON MsgOut where
+    toEncoding = genericToEncoding aesonOpts
+
+daedalusIPC :: Int -> IO ()
+daedalusIPC port = withNodeChannel (pure . msg) action >>= \case
+    Right act -> do
+        sayErr "[INFO] Daedalus IPC server starting"
+        act
+    Left NodeChannelDisabled -> do
+        sayErr "[INFO] Daedalus IPC is not enabled"
+        sleep
```

Comment:
Why sleeping here? What would be the consequence of returning `()` ? Do we expect `daedalusIPC` to never return? If so, I'd rather have it mentioned in a function comment to explain this behavior.

Reply:
Yes, it should have been commented. This is the error handling I want:
1. If set up is unsuccessful, return immediately, causing the cardano-wallet process to exit. Daedalus does not want zombie cardano-wallet processes hanging around that it can't communicate with.
2. If set up is successful, never return until Daedalus exits.
3. If not running within a nodejs channel, never return (simplifies calling code)

Reply:
Okay, so we really assume that any termination is an error, hence the sleep when there's no node.js channel (not an error so to speak).